pkg/lib/registry/registry.go

```diff
@@ -523,3 +449,59 @@ func isDeprecated(ctx context.Context, q *sqlite.SQLQuerier, bundle registry.Bun
 	}
 	return false, nil
 }
+
+func expectedGraphBundles(imagesToAdd []*registry.Bundle, graphLoader registry.GraphLoader, overwrite bool) (map[string]*registry.Package, error) {
+	expectedBundles := map[string]*registry.Package{}
+	for _, bundle := range imagesToAdd {
+		version, err := bundle.Version()
+		if err != nil {
+			return nil, err
+		}
+		newBundleKey := registry.BundleKey{
+			BundlePath: bundle.BundleImage,
+			Version:    version,
+			CsvName:    bundle.Name,
+		}
+		var pkg *registry.Package
+		var ok bool
+		if pkg, ok = expectedBundles[bundle.Package]; !ok {
+			var err error
+			if pkg, err = graphLoader.Generate(bundle.Package); err != nil {
+				if err != registry.ErrPackageNotInDatabase {
+					return nil, err
+				}
+				pkg = &registry.Package{
+					Name:     bundle.Package,
+					Channels: map[string]registry.Channel{},
+				}
+			}
+		}
+		for c, channelEntries := range pkg.Channels {
+			for oldBundle := range channelEntries.Nodes {
+				if oldBundle.CsvName == bundle.Name {
+					if overwrite {
+						delete(pkg.Channels[c].Nodes, oldBundle)
+						if len(pkg.Channels[c].Nodes) == 0 {
+							delete(pkg.Channels, c)
+						}
+					} else {
+						return nil, registry.BundleImageAlreadyAddedErr{ErrorString: fmt.Sprintf("Bundle %s already exists", bundle.BundleImage)}
+					}
+				}
+			}
+		}
+		for _, c := range bundle.Channels {
+			if _, ok := pkg.Channels[c]; !ok {
+				pkg.Channels[c] = registry.Channel{
+					Nodes: map[registry.BundleKey]map[registry.BundleKey]struct{}{},
+				}
+			}
+			// This can miss out on some channels, when a new bundle has channels that the one it replaces does not.
+			// eg: When bundle A in channel A replaces bundle B in channel B is added, bundle B is also added to channel A
+			// but it is only expected to be in channel B, presence in channel A will be ignored
```

Comment:
It will still upgrade to bundle A. We wouldn't verify the bundle's presence on that specific channel though, since channel A would be newly added to bundle B when bundle A is added.